proposals/mainnet_migration.md

```diff
@@ -0,0 +1,540 @@
+    genesisDataObj = {}
+    genesisDataObj.initRounds = DPOS_INIT_ROUNDS
+    r = ceiling((a.header.height - HEIGHT_PREVIOUS_SNAPSHOT_BLOCK) / ROUND_LENGTH)
+    let topDelegates be the set of the top 101 non-banned delegate accounts by delegate weight at the end of round r-2
+    initDelegates = [account.address for account in topDelegates]
+    sort initDelegates in lexicographical order
+    genesisDataObj.initDelegates = initDelegates
+    return genesisDataObj
+```
+
+## Rationale
+
+### Keeping Lisk Core v3 Block History
+
+The decision to discard the block history of Lisk Mainnet at the hard fork from Lisk Core v2 to Lisk Core v3 was considered as rather disadvantageous. One reason is that the transaction history of an account can not be retrieved anymore from a node running Lisk Core v3\. Therefore, the block history of blocks created with Lisk Core v3 shall be kept on Lisk Core v4 nodes.
+
+#### Getting the Lisk Core v3 Block History
+
+In order to keep the implementation of Lisk Core v4 as clean as possible, Lisk Core v4 cannot process Lisk Core v3 blocks, but only ensure their integrity. This is, however, sufficient for maintaining the Lisk Core v3 block history and guaranteeing the immutability of this block history.
+
+Once Lisk Core v4 processed the snapshot block, it can fetch the whole Lisk Core v3 block history, where this is done from the highest to the lowest height. This allows that each block can be validated by simply checking if its block ID matches with the `previousBlockID` property of the child block, and if the payload matches with the `transactionRoot` property. Thus, almost no protocol rules related to Lisk Core v3 must be implemented which keeps the implementation clean. The Lisk Core v3 history can be fetched from a database created by Lisk Core v3 on the same computer or from the peer-to-peer network. Note that the first Lisk Core v3 block - the snapshot block of the previous migration - poses an exception as this one is either [downloaded from a server](#fetching-snapshot-blocks) or fetched from a database created by Lisk Core v3 on the same computer.
+
+### Fetching Snapshot Blocks
+
+The snapshot block created for this hard fork as well as for the [one used for the previous hard fork](https://lisk.observer/block/abdea6dfe5941922d958d1f1dc1d2f97249723b8852325334e55b73e92584191) are not shared via peer-to-peer due to their large size. Instead, they are either downloaded from a server or must be found locally.
+
+#### Initial Version of Lisk Core v4
+
+The initial version of Lisk Core v4 can not download the snapshot block of this hard fork from a server because it could not validate the block. Instead, it must find the snapshot block locally on the same computer. This requires that the snapshot block must be created using a migrator tool on the same node, which in turn requires that the node runs the latest patched version of Lisk Core v3 until the block at height `HEIGHT_SNAPSHOT` is final.
```

Comment:
```suggestion
The initial version of Lisk Core v4 can not download the snapshot block for this hard fork from a server because it could not validate it. Instead, it must find the snapshot block locally on the same computer. This requires that the snapshot block must be created using a migrator tool on the same node, which in turn requires that the node runs the latest patched version of Lisk Core v3 until the block at height `HEIGHT_SNAPSHOT` is final.
```
also please consider "computer" vs "machine"

Reply:
Changed "computer" to "machine" everywhere.